### DISCUSSION TITLE
注入dll的相关问题

### DISCUSSION THREAD

**Discussion by ttttupup** [ttttupup/wxhelper#65] [General]
### 注入工具
1.可以使用项目自带的命令行工具和图形界面工具，**推荐使用命令行工具**，图形工具有mfc的依赖，可能缺少一些dll。
可以使用你喜欢的任意的注入工具注入dll。
命令行工具的使用方法：
ConsoleInject.exe  [-iumpP] :
      -i  注入程序名   -p 注入dll路径     -I 注入程序的pid
     -u 卸载程序名   -d 卸载dll名称
     -m pid  关闭微信互斥体，多开微信
     -P port 指定http端口，需要使用 specify-port 分支的生成的dll
    //注入 dll
    ConsoleInject.exe  -i demo.exe -p E:\\yourDllFullPath.dll
    //卸载 dll
    ConsoleInject.exe  -u demo.exe -d  testInject.dll
    //微信多开 
    ConsoleInject.exe  -m 1222
    // 注入并指定http端口，该命令需要搭配带有specify-port 后缀的dll使用。
    ConsoleInject.exe  -i demo.exe -p E:\\yourDllFullPath.dll -P  18888
2. 3.9.2.23-V6之后的版本支持自定义端口，在微信目录新建config.ini配置文件，内容如下：
  [config]
  port=19088
然后直接使用注入命令时，不需要-P参数指定端口：
tasklist | findstr WeChat.exe    //获取pid  
ConsoleInject.exe  -I  [wechat.exe pid ] -p E:\\yourDllFullPath.dll

### 检查dll是否注入成功
Listdlls.exe   pid   
注意Listdlls.exe 需要自己安装，或者使用其他dll查询工具。
如果没有wxhelper.dll,则说明注入失败，可尝试其他注入工具。
**### 3.9.2.23 release版本会隐藏dll，该方法则查看不到，debug版本可以查看**

### 检查http服务是否正常启动：
tasklist | findstr "WeChat.exe" 
netstat -ano | findstr "your wechat pid"  
如果没有你启动的端口，那么就是http服务启动失败。检查端口是否被占用。
### 如果注入成功，但是http没启动成功，且端口也未被占用，可以检查一下wxhelper.dll所依赖的dll，在本机上是否齐全，
可以使用这个工具或者vs的工具检查一下dll依赖
http://www.dependencywalker.com/
补齐相关依赖重试。特别是ws2_32.dll  wsock32.dll 。
如果还不能正常使用，检查一下vcruntime,到官网重新下载安装
https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170

**Comment by ttttupup** [comment]
https://github.com/nefarius/Injector 这个注入工具比较通用，可以使用这个注入。

**Reply by xushuyi888** [reply]
注入路径不能是中文字符

**Reply by Xxm-YX** [reply]
您好想请教一下， 我使用这个工具注入报Call to LoadLibraryW in remote process failed. 请问有办法解决么

**Reply by Xxm-YX** [reply]
![image](https://github.com/ttttupup/wxhelper/assets/61789235/efee7634-0686-4edb-8117-aa68d5ab3e97)

**Reply by Xxm-YX** [reply]
解决了要用32位的工具

**Reply by Yuri2** [reply]
同上Injector.exe试了64和86的都报一样的错误，是什么问题？
Call to LoadLibraryW in remote process failed.

**Reply by Yuri2** [reply]
用下面的那个兄弟的工具也一样报同样的错误
![3a5c55834e9455c8afe4f06bc62beea](https://github.com/ttttupup/wxhelper/assets/18521198/b67ad715-9638-45f8-a5bb-1b7d94a02edd)

**Reply by litterGuy** [reply]
尝试启动微信的时候用管理员权限启动 试试呢

**Reply by izjing** [reply]
解决这个问题了吗?

**Reply by muxue2333** [reply]
解决了吗，我也出现了同样的问题

**Reply by HermanL02** [reply]
Call to LoadLibraryW in remote process failed. 看了Injector的repo似乎是32位和64位不兼容的意思，“确保在 32 位进程中注入 32 位 DLL， 对于 64 位进程也是如此”。

**Reply by zdev0x** [reply]
使用管理员启动微信，注入成功了，谢谢

**Reply by zddtpy** [reply]
3.9.2.23-v9版本，在微信安装目录新建了一个config.ini文件，在里面写入了内容，提示注入成功了，但是postman调试还是19088端口。端口没有更改到。
[config]
port=19009

**Reply by zddtpy** [reply]
用-P命令，直接就注入结果为0

**Reply by atorber** [reply]
使用作者推荐的注入工具，“以管理员身份运行”启动微信，注入成功

**Reply by atorber** [reply]
确实可以成功

**Reply by ziXiong** [reply]
给“以管理员身份运行”的提个醒，如果你退出微信，记得还要从windows托盘下彻底退出下！  我以为我退出成功了，以管理员身份重启了，其实没有

**Comment by syd2011** [comment]
楼主，想付费咨询您一点技术上问题，wechat：asuka-mme 不胜感激

**Reply by syd2011** [reply]
或者有没有可能我也可以参与到这个开源项目，例如每次微信更新 查找基址和偏移量，测试，一部分feature的功能，例如hook-msg、hook-img

**Reply by ttttupup** [reply]
参与直接提pr就可以了

**Comment by LeanFly** [comment]
注入dll也可以使用 RemoteDLL这个工具，我使用大佬提到的两个注入工具都搞不定，最后用了RemoteDLL搞定了。
[RemoteDLL](https://github.com/LeanFly/wxhelper/blob/main/tool/DLL%20%E6%B3%A8%E5%85%A5%E7%A7%BB%E9%99%A4%E5%B7%A5%E5%85%B7(RemoteDLL)5.0.x86%E6%B1%89%E5%8C%96%E5%8E%BB%E5%B9%BF%E5%91%8A%E7%89%88.exe)

**Comment by Yuri2** [comment]
注入结果反回：0
什么意思？
用Listdlls.exe pid查到没有进程

**Reply by liubowei2018** [reply]
返回零就是没成功，你看你的进程列表上 的pid 和你注入的一致不

**Reply by Yuri2** [reply]
wxhelper-3.9.2.23-v9\tool\injector>ConsoleInject.exe -i WeChat.exe -p wxhelper.dll
dll inject successdll path : w dll path : 10784  注入结果：1


3.9.2.23版本
现在这样子，用测试是否登录接口没有数据反回
https://127.0.0.1:19088/api/?type=8

**Reply by Yuri2** [reply]
C:\Users\tttt\Desktop\5656\wxhelper-3.9.2.23-v9\tool\injector>listdlls.exe pid

用listdlls.exe pid命令测试显示没有注入

Listdlls v3.2 - Listdlls
Copyright (C) 1997-2016 Mark Russinovich
Sysinternals

No matching processes were found.

**Reply by ttttupup** [reply]
http没启动的，可以查一下dll依赖，补齐相关依赖

**Reply by linju324** [reply]
微信版本是3.9.5.8，使用的是对应版本的dll
我这也是使用tools下面ConsoleInject.exe 显示注入结果1，但是没查到有监听19088，listDll.exe 也没查到wxhelper.dll，查了下依赖都有。
换用这个注入工具https://github.com/nefarius/Injector，返回
General Error:
Could not create thread in remote process.

能帮忙看下如何解决吗？

**Reply by ttttupup** [reply]
这个应该是缺少dll，补齐ws_32相关的dll就行了

**Reply by timersun** [reply]
对于一个小白来说，缺少的ws_32相关的dll如何知道都有什么 怎么补齐？

**Reply by ttttupup** [reply]
怎么查，上面已经说了。哪里下载？一个是官方有提供，另一个是直接搜索有网站提供，再一个就是找一台有这个dll的机器直接复制

**Reply by serviceqhd** [reply]
我的电脑dll也不缺失去  搜索system32/里面也是有ws_32  这些dll全有  编译完成的dll依赖库都有 但是每次都是注入成功了 请求接口都没反应 19088也没有起来   版本是3.9.2.23 就是主分支main

**Comment by caiyakang** [comment]
请问如何卸载DLL？无法卸载DLL

**Comment by timersun** [comment]
好的  我个另外一个注入工具注入了，可能是我工具的问题，楼主👍👍👍👍👍👍👍👍👍

ttttupup ***@***.***> 于2023年10月20日周五 10:12写道：

**Reply by qianqian1530** [reply]
啥注入器啊 ..分享一下被...

**Comment by timersun** [comment]
https://github.com/nefarius/Injector

qianqian1530 ***@***.***> 于2023年12月6日周三 15:14写道：

**Comment by HermanL02** [comment]
总结，用大佬提供的https://github.com/nefarius/Injector完成了注入，测试版本3.9.8.25, 3.9.8.25 v2, 
1. 选择x64版本的注入器, 注入x64的微信程序
2. 之前版本微信先完全卸载后，重启安装旧版本微信
3. Injector.exe --process-name WeChat.exe -i D:\wxhelper3982501.dll
针对楼上的一些误解，就是这个WeChat.exe指的是正在运行的程序，不需要把任何东西放到微信的安装目录

**Reply by cynen** [reply]
DLL注入指的是向运行中的其他进程(程序WeChat.exe)强制插入特定的DLL(wxhelper.dll)文件.  
> 补一句: 正经的程序谁会去操作其他的进程空间呢？你懂得. O(∩_∩)O哈哈~

**Comment by xiaoming6617** [comment]
使用的是3.9.8.25版本分支，使用版本里的injector.exe，显示注入成功
cDllPath:D:\wxhelper.dll wsPath:D
dll inject success
dll path : D:\wxhelper.dll
dll pid : 34924
 注入结果：1

实际Listdlls.exe 34924没有找到wxhelper.dll，这个要怎么查？？？

更多信息如下：
netstat -ano | findstr  "34924"的结果是
  TCP    0.0.0.0:19088          0.0.0.0:0              LISTENING       34924
  TCP    127.0.0.1:8680         0.0.0.0:0              LISTENING       34924
  TCP    172.17.186.194:58388   14.116.242.216:80      ESTABLISHED     34924

（采用https://github.com/nefarius/Injector的注入器也是显示成功，但是listdll找不到这个库）

**Reply by ttttupup** [reply]
配置文件里可以配置是否隐藏dll，如果隐藏了。listdll就查不到

**Reply by xiaoming6617** [reply]
这个问题原因确实是因为默认配置HiddenDll是1导致，改为0即可以看到，学习了

**Reply by yangchunyong** [reply]
我的也是，看起来成功了
TCP    0.0.0.0:19088          0.0.0.0:0              LISTENING       35980
  TCP    127.0.0.1:8680         0.0.0.0:0              LISTENING       35980
  TCP    127.0.0.1:19088        127.0.0.1:51300        ESTABLISHED     35980
  TCP    192.168.2.17:53425     117.89.177.97:8080     ESTABLISHED     35980
  但是从check_login开始所有的接口都是返回：
{"code":200,"data":null,"msg":"the url is not supported"}

{"code":200,"data":null,"msg":"the url is not supported"}

{"code":200,"data":null,"msg":"the url is not supported"}

{"code":200,"data":null,"msg":"the url is not supported"}
不知道什么原因。用的最新的3.9.8.25 v2

**Reply by dchcrzm** [reply]
有解决这个问题嘛 我的也是 ，接口全部返回{"code":200,"data":null,"msg":"the url is not supported"}

**Reply by dongyi-Meng** [reply]
这个简单，只支持post方式，不支持get。就不能在浏览器直接打开

**Comment by zhaihao118** [comment]
我是windows10系统，微信用超级管理员启动，inject在超级管理员启动的命令行执行，可以成功注入。

另外，注入不成功的，还需要检查下运行时库是否存在

**Reply by yinhuacha869** [reply]
运行时库是指什么？我有一台电脑可以，有一台电脑不可以，不可以的注入时提示
General Error:
Call to LoadLibraryW in remote process failed.

**Comment by ywx1216756599** [comment]
大佬们 我这边多开微信注入dll 都成功了 但是接受到的信息只有一个pid的 这么处理啊

**Comment by dongyi-Meng** [comment]
学了几天，找了个call
用c++调用总是崩溃。有没有懂点的教一下
这个call是获取code的
正常的call断点状态
![Image_1726421035609](https://github.com/user-attachments/assets/31485885-4342-4210-a264-80d7bc95d23a)
vs的代码，参照楼主的sendmsg写的
![Image_1726421025416](https://github.com/user-attachments/assets/692803c0-4a29-479f-9cca-0ba8a08f9c7a)
传递进去后崩溃
![Image_1726421029051](https://github.com/user-attachments/assets/5c229797-f46c-434f-8234-89b54b81a4fc)

**Comment by friendbookok** [comment]
假如想编程解决遇到的一堆的报错，在win10上，或者一些机器上是可能不行的，因为很多人用的是盗版或者裁剪版本的，会缺少很多dll，而且这些缺少的dll用作者提供的路径可能是解决不了的，比如我在win10上就试过作者的以下方法，是无法解决的：
可以使用这个工具或者vs的工具检查一下dll依赖
http://www.dependencywalker.com/
补齐相关依赖重试。特别是ws2_32.dll wsock32.dll 。
如果还不能正常使用，检查一下vcruntime,到官网重新下载安装
https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170。

其实更加好的方案就是狠心下载一个windows sdk.下载自己配合系统的sdk就可以完全解决了。免得定位各种dll缺少问题浪费时间。
下面附上win 10官方sdk链接
https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/
其它的自己通过ai或者google来处理。

当然，假如是想自己玩一下，使用采用https://github.com/nefarius/Injector 链接里的也是ok的，它里面应该是打包了dll进去了，是可以正常运行的。需要注意的是这个injector使用的命令行参数与[ttttupup](https://github.com/ttttupup)大神使用的参数是完全不同的，毕竟是2个不同的大神了。

**Reply by friendbookok** [reply]
![image](https://github.com/user-attachments/assets/79db9451-b110-4ffa-8d5f-0d9ad5bd3813)
这是成功后的postman图，顺便说一下，postman.json路径在作者的以下路径下。
![image](https://github.com/user-attachments/assets/7cc80d4c-b97a-4359-a33b-f3fceb9c46c4)

**Reply by LeafRot** [reply]
请问你用的是什么版本，能获取图片吗？
